packages/block-library/src/table-of-contents/hooks.js

```diff
@@ -0,0 +1,153 @@
+		// relative to sibling blocks.
+		const tocIndex = allBlockClientIds.indexOf( clientId );
+
+		for ( const [
+			blockIndex,
+			blockClientId,
+		] of allBlockClientIds.entries() ) {
+			// If we've reached blocks after the Table of Contents, we've
+			// finished calculating which page the block is on.
+			if ( blockIndex >= tocIndex ) {
+				break;
+			}
+			if ( getBlockName( blockClientId ) === 'core/nextpage' ) {
+				tocPage++;
+			}
+		}
+	}
+
+	const latestHeadings = [];
+
+	/** The page (of a paginated post) a heading will be part of. */
+	let headingPage = 1;
+	let headingPageLink = null;
+
+	// If the core/editor store is available, we can add permalinks to the
+	// generated table of contents.
+	if ( typeof permalink === 'string' ) {
+		headingPageLink = isPaginated
+			? addQueryArgs( permalink, { page: headingPage } )
+			: permalink;
+	}
+
+	for ( const blockClientId of allBlockClientIds ) {
+		const blockName = getBlockName( blockClientId );
+		if ( blockName === 'core/nextpage' ) {
+			headingPage++;
+
+			// If we're only including headings from the current page (of
+			// a paginated post), then exit the loop if we've reached the
+			// pages after the one with the Table of Contents block.
+			if ( onlyIncludeCurrentPage && headingPage > tocPage ) {
+				break;
+			}
+
+			if ( typeof permalink === 'string' ) {
+				headingPageLink = addQueryArgs(
+					removeQueryArgs( permalink, [ 'page' ] ),
+					{ page: headingPage }
+				);
+			}
+		}
+		// If we're including all headings or we've reached headings on
+		// the same page as the Table of Contents block, add them to the
+		// list.
+		else if ( ! onlyIncludeCurrentPage || headingPage === tocPage ) {
+			if ( blockName === 'core/heading' ) {
+				const headingAttributes = getBlockAttributes( blockClientId );
+
+				const canBeLinked =
+					typeof headingPageLink === 'string' &&
+					typeof headingAttributes.anchor === 'string' &&
+					headingAttributes.anchor !== '';
+
+				latestHeadings.push( {
+					// Convert line breaks to spaces, and get rid of HTML tags in the headings.
+					content: stripHTML(
+						headingAttributes.content.replace(
+							/(<br *\/?>)+/g,
+							' '
+						)
+					),
+					level: headingAttributes.level,
+					link: canBeLinked
+						? `${ headingPageLink }#${ headingAttributes.anchor }`
+						: null,
+				} );
+			}
+		}
+	}
+
+	return latestHeadings;
+}
+
+function observeCallback( select, dispatch, clientId ) {
+	const { getBlockAttributes } = select( blockEditorStore );
+	const { updateBlockAttributes, __unstableMarkNextChangeAsNotPersistent } =
+		dispatch( blockEditorStore );
+
+	const headings = getLatestHeadings( select, clientId );
+	const storedHeadings = getBlockAttributes( clientId )?.headings;
+
+	// When undoing changes to an attribute, `getBlockAttributes` returns `null`,
+	// which triggers an infinite loop of updates. The `undefined` check prevents that.
+	if (
+		storedHeadings !== undefined &&
```

Comment:
This condition prevents the `headings` attribute from being set when I create the TOC block from scratch. Then the initial value of the attribute is `undefined`, and the observer hook never updates it.

Maybe the `headings` attribute should have a `[]` default value specified in `block.json`.

How do I trigger the infinite update loop that you're trying to prevent here? I suppose it's triggered by undo. I.e., I insert a new heading and on the undo stack there is _one_ record with _two_ updates: block insertion and TOC attribute update. Then I press <kbd>Ctrl-Z</kbd> and it causes both updates to be undone, in one step. And the observer hook should handle that gracefully. Is that right?

Reply:
Thanks for catching that!

> Maybe the headings attribute should have a [] default value specified in block.json.
Yes. I'm going to add an empty default value.

How to trigger an infinite update loop:

1. Remove the safeguard and build the code.
2. Open a post.
3. Add ToC blocks.
4. Add a few headings.
5. Save post and reload.
6. Make changes to one heading.
7. Undo all changes. The last undo step triggers the loop.

This highlights that the block attribute isn't reverted to an empty state (`undefined`), but `getBlockAttributes` returns `null` for some reason at that point.

Reply:
> How to trigger an infinite update loop:

I found what is going on. When you perform undo, the TOC block is replaced with a new one, with the old attributes, and also with a new `clientId`.

But at the time this `RESET_BLOCKS` action is done and the observer reacts to it, the old block's `Edit` component is still mounted, and the `useObserveHeadings` hook is also still mounted. The effect cleanup will happen a bit later.

What we need to do in the observer loop is to detect that the `clientId` block no longer exists. That should be easy because the `getBlockAttributes` selector returns `null`. You already know that, you wrote it into the comment 🙂 Now we also know what it means. We need to `return` early from the callback when this condition is detected, and do no update.

Reply:
I added an early return in [2c9dc84](https://github.com/WordPress/gutenberg/pull/54094/commits/2c9dc8404866f13bb27f18f6a2e2c03bd0327df9) + comment. Let me know if you want to include more details in the comment.

Reply:
Thanks for the fix, I have nothing to add to the comment.

> When you perform undo, the TOC block is replaced with a new one, with the old attributes, and also with a new `clientId`.

This, however, seems to be another undo quirk (in addition to #41031). In most cases, undo and redo keep the blocks' `clientId`s stable, it just modifies the attributes. But in the specific case, where the "heading typing" is the very first edit after loading the post, the undo (namely the `RESET_BLOCKS` action) regenerates all the affected `clientId`s. Worth investigating.

Reply:
Do you think that's a different issue but more highlighted by the ToC observer? I'm just wondering if this should be a blocker for this refactoring.

As I mentioned in my previous comment, I plan to change the storing mechanism for `headings` in follow-up PR.